### ISSUE TITLE
Enveloped signature: Signature-placeholder not found when not direct child of root

### ISSUE THREAD

**Issue by vboehm** [kislyuk/signxml#32]
Hello,

According to then W3C specification an enveloped signature is
> **Signature, Enveloped**
> The signature is over the XML content that contains the signature as an element. The content provides the root XML document element. Obviously, enveloped signatures must take care not to include their own value in the calculation of the SignatureValue.

There is no need for the signature element to be a direct child of the root element.

So you may use

    <TheInformation>
        <SomeInfo>
            ...
            <Security>
                ...
                <Signature>  
                    ...
                </Signature>  
                ...
            </Security>
            ...
        <\SomeInfo>
        <SomeOtherInfo>
            ...
        <\SomeOtherInfo>
        ...
    </TheInformation>

And there are many implementations which put the _signature_ element **somewhere** in the
XML-tree.

Your implementation (method _\_getPayload\_c14n_) uses

    .
    .
    signature_placeholders = self._findall(self.data, "Signature[@Id='placeholder']")
    .
    .

to find the position of the signature element and _\_findall_ is defined as

    def _findall(self, element, query, namespace="ds"):
        return element.findall(namespace + ":" + query, namespaces=namespaces)

But the element-method _findall_ only searches for direct childs if called with a simple name

    element.findall('ds:Signature')

To find all descendants you have to search with an x-path  expression beginning with './/' 

    element.findall('.//ds:Signature')

I changed the definition of _\_findall_ to

    def _findall(self, element, query, namespace="ds"):
        return element.findall('.//' + namespace + ":" + query, namespaces=namespaces)

and now the Program behaves correctly. 

But since you use your _\_findall_ method in several places, I don't know if I destroyed the other uses of this routine. 
And until now I didn't find the time to analyze, if the non-recursive search occurs in other places of your program (I think your _verify_ function does also not work for the above example).

regards  Volker

**Comment by jmindek** [comment]
I am also receiving <ds:Signature> tags that are not direct children of <samlp:Response>.

I have created a branch which change the _find and _findall method signature to:
```    
def _find(self, element, query, require=True, namespace="ds", anywhere=False):
```
and
``` 
def _findall(self, element, query, namespace="ds", anywhere=False):
```
Inside the function body is an if check on `anywhere`. When True, the '://' is added to the query.

I do not have permission to branch and push to this repo so I am not able to contribute this change as of this comment.

**Comment by kislyuk** [comment]
Thank you for your contributions, @vboehm and @jmindek. I like @jmindek's solution and was thinking along similar lines. I agree with the change in general.

Can you both please open pull requests? @jmindek, you can open a pull request by simply editing any file with the pencil icon - it will automatically fork the repo into your account and let you open a branch for us to review.

I want you both to open PRs so we can select the best implementation ideas from both.

Thank you.

**Comment by jmindek** [comment]
Is there any chance we can get this merged soon? I have an immediate need for it.

**Comment by kislyuk** [comment]
Sorry about the delay - will take a look and see what I can do.

**Comment by kislyuk** [comment]
Fixed from #33, released in v0.4.5. Please test.
